### PR TITLE
Introduce JsPlugin for configuration and run tests on NodeJS

### PR DESCRIPTION
**Pull request by CherryPerry**
`jsTest` task does not run tests.

To run test we need to install `mocha` (probably any JS test engine is fine) and launch it manually through `node mocha reaktive-test.js` call. 

Fixes https://github.com/badoo/Reaktive/issues/53

### CODE REVIEW

**Comment by CherryPerry** [comment]
Works fine: https://travis-ci.org/badoo/Reaktive/jobs/534482640#L1427

File: buildSrc/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+    `kotlin-dsl`
+}
+
+repositories {
+    google()
+    jcenter()
+}
+
+dependencies {
+    implementation("com.moowork.gradle:gradle-node-plugin:1.3.1")
```
Comment (arkivanov) [inline buildSrc/build.gradle.kts]: Shouldn't it be ```implementation `xxx` ```?

File: reaktive/build.gradle [hunk 2]
```diff
@@ -56,29 +54,19 @@ if (target.isCommon() || target.isMeta()) {
...
+                    implementation 'org.jetbrains.kotlin:kotlin-test-junit'
                 }
             }
 
-            jsTest {
-                dependencies {
-                    implementation 'org.jetbrains.kotlin:kotlin-test-js'
-                }
-            }
+            jvmMain.dependsOn jvmCommonMain
```
Comment (arkivanov) [inline reaktive/build.gradle]: Let's group `xxxMain` and `xxxTest` together?

**Comment by arkivanov** [comment]
> Works fine: https://travis-ci.org/badoo/Reaktive/jobs/534482640#L1427

So JS tests run automatically on CI? How exactly we can run them manually?

File: buildSrc/build.gradle.kts (hunk 1 above)
Comment (CherryPerry) [reply buildSrc/build.gradle.kts]: It is kotlin script, not groovy 😎

**Comment by CherryPerry** [comment]
Just run `./gradlew reaktive:runMocha`

**Comment by arkivanov** [comment]
> Just run `./gradlew reaktive:runMocha`

Are they running automatically on CI? Am I right that `jsTest` task does not actually run JS tests? What about `check` task?

**Comment by CherryPerry** [comment]
This task was added to task invocation list for TravisCI in `.travis.yml`.

`jsTest` does only compilation, no tests are run.

I don't want to make `check` tasks depend on `runMocha` task for now. It requires additional NodeJS dependency, which will be downloaded on invocation.

**Comment by arkivanov** [comment]
> This task was added to task invocation list for TravisCI in `.travis.yml`.
> 
> `jsTest` does only compilation, no tests are run.
> 
> I don't want to make `check` tasks depend on `runMocha` task for now. It requires additional NodeJS dependency, which will be downloaded on invocation.

That makes sense

**Comment by lukaville** [comment]
> This task was added to task invocation list for TravisCI in `.travis.yml`.
> 
> `jsTest` does only compilation, no tests are run.
> 
> I don't want to make `check` tasks depend on `runMocha` task for now. It requires additional NodeJS dependency, which will be downloaded on invocation.

What's the problem with NodeJS dependency? I think we should add this task to `check`

File: buildSrc/src/main/kotlin/JsPlugin.kt [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        @get:OutputFile
+        abstract val output: Property<File>
+
+        init {
+            output.set(input.map { project.file("node_modules/${it.name}") })
+        }
+
+        @TaskAction
+        open fun populate() {
+            input.get().copyTo(target = output.get(), overwrite = true)
```
Comment (lukaville) [inline buildSrc/src/main/kotlin/JsPlugin.kt]: Can we use [Copy](https://docs.gradle.org/current/dsl/org.gradle.api.tasks.Copy.html) task for this?

File: buildSrc/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    implementation("org.jetbrains.kotlin:kotlin-gradle-plugin:1.3.31")
```
Comment (lukaville) [inline buildSrc/build.gradle.kts]: let's use `ext.kotlin_version` variable?

File: buildSrc/src/main/kotlin/JsPlugin.kt [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        }
+
+        target.pluginManager.apply(NodePlugin::class.java)
+        target.extensions.configure(NodeExtension::class.java) {
+            download = true
+            version = "12.2.0"
+        }
+
+        val dependenciesTaskProvider = target.tasks.register("installJsTestDependencies", NpmTask::class.java) {
+            setArgs(listOf("install", "kotlin@1.3.31", "kotlin-test@1.3.31", "mocha@6.1.4"))
```
Comment (lukaville) [inline buildSrc/src/main/kotlin/JsPlugin.kt]: let's use `ext.kotlin_version` variable?

Comment (CherryPerry) [reply buildSrc/src/main/kotlin/JsPlugin.kt]: Extracted to `Versions.kotlin`, ext `build.gradle` is not available in `buildSrc`.

File: buildSrc/build.gradle.kts (hunk 4 above)
Comment (CherryPerry) [reply buildSrc/build.gradle.kts]: Extracted to `Versions.kotlin`, ext `build.gradle` is not available in `buildSrc`.

**Comment by arkivanov** [comment]
@CherryPerry could you please merge master?

**Comment by CherryPerry** [comment]
@arkivanov Need to check something else, will do it tomorrow. If you want to merge `jvmJsCommon` source set, do it first, I will resolve conflicts in Gradle files and update my branch.

**Comment by arkivanov** [comment]
> @arkivanov Need to check something else, will do it tomorrow. If you want to merge `jvmJsCommon` source set, do it first, I will resolve conflicts in Gradle files and update my branch.

I can't merge, waiting for approval from @mk-jones :-)

**Pull request merged**